crates/precompile/src/bn128/arkworks.rs

```diff
@@ -0,0 +1,213 @@
+#[inline]
+fn new_g1_point(px: Fq, py: Fq) -> Result<G1Affine, PrecompileError> {
+    if px.is_zero() && py.is_zero() {
+        Ok(G1Affine::zero())
+    } else {
+        let point = G1Affine::new_unchecked(px, py);
+        if !point.is_on_curve() || !point.is_in_correct_subgroup_assuming_on_curve() {
+            return Err(PrecompileError::Bn128AffineGFailedToCreate);
+        }
+        Ok(point)
+    }
+}
+
+/// Creates a new `G2` point from the given Fq2 coordinates.
+///
+/// G2 points in BN128 are defined over a quadratic extension field Fq2.
+/// This function takes two Fq2 elements representing the x and y coordinates
+/// and creates a G2 point.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG2` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG2` to `G2`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g2_point(x: Fq2, y: Fq2) -> Result<G2Affine, PrecompileError> {
+    let point = if x.is_zero() && y.is_zero() {
+        G2Affine::zero()
+    } else {
+        let point = G2Affine::new_unchecked(x, y);
+        if !point.is_on_curve() || !point.is_in_correct_subgroup_assuming_on_curve() {
+            return Err(PrecompileError::Bn128AffineGFailedToCreate);
+        }
+        point
+    };
+
+    Ok(point)
+}
+
+/// Reads a G1 point from the input slice.
+///
+/// Parses a G1 point from a byte slice by reading two consecutive field elements
+/// representing the x and y coordinates.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 64 bytes long.
+#[inline]
+pub(super) fn read_g1_point(input: &[u8]) -> Result<G1Affine, PrecompileError> {
+    let px = read_fq(&input[0..FQ_LEN])?;
+    let py = read_fq(&input[FQ_LEN..2 * FQ_LEN])?;
+    new_g1_point(px, py)
+}
+
+/// Encodes a G1 point into a byte array.
+///
+/// Converts a G1 point in Jacobian coordinates to affine coordinates and
+/// serializes the x and y coordinates as big-endian byte arrays.
+///
+/// Note: If the point is the point at infinity, this function returns
+/// all zeroes.
+#[inline]
+pub(super) fn encode_g1_point(point: G1Affine) -> [u8; G1_LEN] {
+    let mut output = [0u8; G1_LEN];
+
+    if let Some((x, y)) = point.xy() {
+        // TODO: check if serialize_uncompressed ever writes less than FQ_LEN
+        let mut x_bytes = [0u8; FQ_LEN];
+        x.serialize_uncompressed(&mut x_bytes[..])
+            .expect("Failed to serialize x coordinate");
+
+        let mut y_bytes = [0u8; FQ_LEN];
+        y.serialize_uncompressed(&mut y_bytes[..])
+            .expect("Failed to serialize x coordinate");
+
+        // Convert to big endian by reversing the bytes.
+        x_bytes.reverse();
+        y_bytes.reverse();
+
+        // Place x in the first half, y in the second half.
+        output[0..FQ_LEN].copy_from_slice(&x_bytes);
+        output[FQ_LEN..(FQ_LEN * 2)].copy_from_slice(&y_bytes);
```

Comment:
We could encode x into x_bytes, then write to output and then re-use x_bytes -- though this stack allocation is fast, so its not necessary imo